drivers/gpio/gpio_mchp_xec_v2.c

```diff
@@ -368,6 +378,128 @@ static void gpio_gpio_xec_port_isr(const struct device *dev)
+ */
+
+/**
+ * @brief Called by PINCTRL, configures drive and slew rate
+ * @param dev pointer to GPIO port device
+ * @param pin pin bit position in GPIO port
+ * @param cfg PINCTRL encoded pin properties:
+ *       Pulls, output buffer type, slew rate, drive strength
+ */
+
+static void config_drive_slew(const struct device *dev,
+			      gpio_pin_t pin, uint32_t conf)
+{
+	uintptr_t pcr2_addr = pin_ctrl2_addr(dev, pin);
+	uint32_t slew = conf &
+		(MCHP_XEC_OSPEEDR_MASK << MCHP_XEC_OSPEEDR_POS);
+	uint32_t drvstr = conf &
+		(MCHP_XEC_ODRVSTR_MASK << MCHP_XEC_ODRVSTR_POS);
+	uint32_t val = 0;
+	uint32_t mask = 0;
+
+	if (slew != MCHP_XEC_OSPEEDR_NO_CHG) {
+		mask |= MCHP_GPIO_CTRL2_SLEW_MASK;
+		if (slew == MCHP_XEC_OSPEEDR_FAST) {
+			val |= MCHP_GPIO_CTRL2_SLEW_FAST;
+		}
+	}
+	if (drvstr != MCHP_XEC_ODRVSTR_NO_CHG) {
+		mask |= MCHP_GPIO_CTRL2_DRV_STR_MASK;
+		val |= (drvstr << MCHP_GPIO_CTRL2_DRV_STR_POS);
+	}
+
+	if (!mask) {
+		return;
+	}
+
+	xec_mask_write32(pcr2_addr, mask, val);
+}
+
+static void config_alt_func(const struct device *dev,
+			    gpio_pin_t pin, uint32_t altf)
+{
+	uintptr_t pcr1_addr = pin_ctrl_addr(dev, pin);
+	uint32_t muxval = (uint32_t)((altf & MCHP_GPIO_CTRL_MUX_MASK0) <<
+				     MCHP_GPIO_CTRL_MUX_POS);
+
+	xec_mask_write32(pcr1_addr, MCHP_GPIO_CTRL_MUX_MASK, muxval);
+}
+
+/**
+ * @brief Called by PINCTRL, configures pin
+ * @param dev pointer to GPIO port device
+ * @param pin pin bit position in GPIO port
+ * @param cfg PINCTRL encoded pin properties:
+ *       Pulls, output buffer type, slew rate, drive strength
+ * @param altf = alternate function where 0 = GPIO
+ *
+ * @retval 0 on success.
+ * @retval -EINVAL if pin is not valid.
+ */
+int gpio_mchp_xec_configure(const struct device *dev, gpio_pin_t pin,
+			    uint32_t conf, uint32_t altf)
+{
+	gpio_flags_t flags = 0;
+	uint32_t temp = 0;
+	int ret = 0;
+
+	temp = (conf & MCHP_XEC_PUPDR_MASK) >> MCHP_XEC_PUPDR_POS;
+	switch (temp) {
+	case MCHP_XEC_PULL_UP:
+		flags |= GPIO_PULL_UP;
+		break;
+	case MCHP_XEC_PULL_DOWN:
+		flags |= GPIO_PULL_DOWN;
+		break;
+	case MCHP_XEC_REPEATER:
+		flags |= (GPIO_PULL_UP | GPIO_PULL_DOWN);
+		break;
+	default:
+		break;
+	}
+
+	if ((conf >> MCHP_XEC_OTYPER_POS) & MCHP_XEC_OTYPER_MASK) {
+		flags |= GPIO_LINE_OPEN_DRAIN;
+	}
+
+	temp = (conf >> MCHP_XEC_OVAL_POS) & MCHP_XEC_OVAL_MASK;
+	if (temp) {
+		flags |= GPIO_OUTPUT;
+		if (temp == MCHP_XEC_OVAL_DRV_HIGH) {
+			flags |= GPIO_OUTPUT_INIT_HIGH;
+		} else {
+			flags |= GPIO_OUTPUT_INIT_LOW;
+		}
+	} else {
+		flags |= GPIO_INPUT;
+	}
+
+	if (conf & BIT(MCHP_XEC_PIN_LOW_POWER_POS)) {
+		flags = GPIO_DISCONNECTED;
+	}
+
+	ret = gpio_xec_configure(dev, pin, flags);
+	if (ret) {
+		return ret;
+	}
+
+	if (flags != GPIO_DISCONNECTED) {
+		/* program slew rate and drive strength */
+		config_drive_slew(dev, pin, conf);
+
+		/* program alternate function */
+		config_alt_func(dev, pin, altf);
+	}
+
+	return 0;
+}
+
```

Comment:
This should all be moved to the pinctrl driver (which is where this is used, really). As an example, having code that configures alternate function in a GPIO driver doesn't make much sense. And this also creates a dependency between pinctrl & GPIO for no reason. STM32 does something similar, but mostly for historical reasons, and should probably be fixed. GD32 platform (similar/cloned hardware) has independent GPIO/pinctrl drivers.

Reply:
@gmarull 
Yes will move the three functions marked to pinctrl driver.

Could you review other all other commits in this PR and provide your comments? We like to get this PR approved by you in the next iteration of commits.

Reply:
Moved these functions to pinctrl driver.